plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/client/BackpressureRestClient.java

```diff
@@ -0,0 +1,120 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.elasticsearch.client;
+
+import com.google.common.base.Stopwatch;
+import io.airlift.log.Logger;
+import io.airlift.stats.TimeStat;
+import io.trino.plugin.elasticsearch.ElasticsearchConfig;
+import net.jodah.failsafe.Failsafe;
+import net.jodah.failsafe.FailsafeException;
+import net.jodah.failsafe.RetryPolicy;
+import net.jodah.failsafe.event.ExecutionCompletedEvent;
+import net.jodah.failsafe.function.CheckedSupplier;
+import org.apache.http.Header;
+import org.apache.http.HttpEntity;
+import org.apache.http.HttpHost;
+import org.elasticsearch.client.Response;
+import org.elasticsearch.client.ResponseException;
+import org.elasticsearch.client.RestClient;
+import org.elasticsearch.rest.RestStatus;
+
+import java.io.IOException;
+import java.util.Map;
+
+import static com.google.common.base.Throwables.throwIfInstanceOf;
+import static com.google.common.base.Throwables.throwIfUnchecked;
+import static java.time.temporal.ChronoUnit.MILLIS;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class BackpressureRestClient
+{
+    private static final Logger log = Logger.get(BackpressureRestClient.class);
+
+    private final RestClient delegate;
+    private final RetryPolicy<Response> retryPolicy;
+    private final TimeStat retryStats;
+    private final ThreadLocal<Stopwatch> stopwatch = ThreadLocal.withInitial(Stopwatch::createUnstarted);
+
+    public BackpressureRestClient(RestClient delegate, ElasticsearchConfig config, TimeStat retryStats)
+    {
+        this.delegate = requireNonNull(delegate, "restClient is null");
+        this.retryStats = requireNonNull(retryStats, "retryStats is null");
+        requireNonNull(config, "config is null");
+        retryPolicy = new RetryPolicy<Response>()
+                .withMaxAttempts(-1)
+                .withMaxDuration(java.time.Duration.ofMillis(config.getMaxRetryTime().toMillis()))
+                .withBackoff(config.getMinRetryTime().toMillis(), config.getMaxRetryTime().toMillis(), MILLIS)
+                .withJitter(0.125)
+                .handleIf(BackpressureRestClient::isBackpressure)
+                .onFailedAttempt(executionAttemptedEvent -> {
+                    log.debug("REST attempt failed: %s", executionAttemptedEvent.getLastFailure());
+                    stopwatch.get().start();
+                })
+                .onRetry(executionAttemptedEvent -> stopwatch.get().stop())
```

Comment:
You should account retry time too if we need to `Clearly report any extra delay introduced by retry`.
Please make sure that `stopwath` returns 0 when there is no retry (preferably test)